<hr style="border: 6px solid#003262;" />

<div align="center">
    <img src= "/assets/content/datax_logos/DataX_blue_wide_logo.png" align="center" width="100%">
</div>


<br>

# **DATA-X:** <br>  m410 - SHALLOW NEURAL NETWORKS; AN INTRODUCTION TO TENSORFLOW

<br>


**Author List (in no particular order):** [Elias Castro Hernandez](https://www.linkedin.com/in/ehcastroh/), [Ikhlaq Sidhu](https://ikhlaq-sidhu.com/), [Debbie Yuen](http://www.debbiecyuen.me/), and [Alexander Fred-Ojala](https://www.linkedin.com/in/alexanderfo/) 

**About (TL/DR):** Tensorflow (TF) is an open-source library used for dataflow, differentiable programming, symbolic math ,and machine learning applications such as deep learning neural networks. TF's flexible architecture allows for easy deployment across varied processing platforms. 

**Learning Goal(s):** This notebook covers advanced topics in machine learning. However, it does not require any prior knowledge in machine learning. The goal of this notebook is to teach a user how to deploy a TF model, as well as to provide the user guidance on how to tackle the more nuanced topics. 

**Associated Materials:** To ease the learning curve, we encourage the user of this notebook to view the resources section on the main JupyterLab, and/or review the [Data-X Fundamentals](https://github.com/scetx/datax/tree/master/01-data-x-fundamentals) repo.

**Keywords (Tags):** tensorflow, tensor-flow, tensorflow-tutorial, deep-learning, deep-learning-with-python, neural-networks, data-x, uc-berkeley-engineering 

**Prerequisite Knowledge:** (1) Python, (2) NumPy, (3) Pandas, (4) Linear Algebra, (5) Bash 

**Target User:** Data scientists, applied machine learning engineers, and developers

**Copyright:** Content curation has been used to expedite the creation of the following learning materials. Credit and copyright belong to the content creators used in facilitating this content. Please support the creators of the resources used by frequenting their sites, and social media.


<hr style="border: 4px solid#003262;" />

<a name='Part_table_contents' id="Part_table_contents"></a>



#### CONTENTS

> #### [PART 0: ABOUT AND MOTIVATION](#Part_0)
> #### [PART 1: TENSORS AND OPERATIONS](#Part_1)
> #### [PART 2: TENSORFLOW GRAPHS, AND EXECUTIONS](#Part_2)
> #### [PART 3 (OPTIONAL): TENSORFLOW LINEAR REGRESSION](#Part_3)
> #### [PART 4: WRAP UP AND NEXT STEPS](#Part_4)

#### APPENDIX

> #### [PREREQUISITE KNOWLEDGE AND REQUIREMENTS](#Appendix_1)
> #### [REFERENCES AND ADDITIONAL RESOURCES](#Appendix_2)

<br>


<a id='Part_0'></a>

<hr style="border: 2px solid#003262;" />

#### PART 0

## **ABOUT** AND **MOTIVATION** NEED VIDEO FROM DEBBIE

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
  <a href="">
    <img src="assets/content/images/tensorflow_thumbnail_play-01.png" align="center" width="50%" padding="10px"/>
  </a><br>
</div>

<a id='Part_1'></a>

<hr style="border: 2px solid#003262;" />

#### PART 1

## **TENSORS** AND **OPERATIONS**


<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/tensorflow_thumbnail-01.png" align="center" width="40%" padding="10"><br>
    <br>
    
</div>

<br>

<br>

[**TensorFlow**](https://www.tensorflow.org/) is a cross-platform, end-to-end, open source, platform for efficiently training and deploying machine learning models. TensorFlow offers multiple levels of abstractions, which means you can use high-level API's such as [**Keras**](https://keras.io/) or [**Ludwig**](https://ludwig-ai.github.io/ludwig-docs/index.html) to make things a bit simpler, or even to set up a [**Distribution Strategy**](https://www.tensorflow.org/guide/distributed_training) API on different hardware configurions without having to change the model defintion.

<br>

<strong style="color:red">KEY CONSIDERATION:</strong> Some of the following content may be written for machines running on Linux or Mac operating systems. If you are working on a Windows machine, you will need to enable the Linux Bash Shell, or adjust Shell commands to PowerShell syntax. A tutorial on how to enable the Linux Bash Shell on Windows 10 can be found [here](https://youtu.be/xzgwDbe7foQ).


#### CONTENTS:

> [PART 1.0: TENSORFLOW INSTALLATION NOTES](#Part_1_0)<br>
> [PART 1.1: TENSORFLOW SETUP](#Part_1_1)<br>
> [PART 1.2: TENSORBOARD SETUP](#Part_1_2)<br>
> [PART 1.3: TENSORFLOW TENSORS](#Part_1_3)<br>
> [PART 1.4: TENSORFLOW OPERATIONS](#Part_1_4)<br>
> [PART 1.5: TENSORFLOW COMPUTATION GRAPHS](#Part_1_5)

<a id='Part_1_0'></a>

<hr style="border: 1px solid#003262;" />

#### PART 1.0: TENSORFLOW INSTALLATION NOTES

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/tf_logo_social.png" align="center" width="30%" padding="0px"><br>
    <br>
</div>

<br>

#### **1.0.1 Install Python Development Environment**

___



#### **Install/Update pip3**

https://pip.pypa.io/en/stable/installing/

```bash
    # TensorFlow requires pip version >= 19.0
    $ pip install --upgrade pip 
```

<br>

#### **Install/Update Python 3**

https://www.python.org/downloads/

```bash
    # TensorFlow requires Python 3.5-3.8 
    $ sudo apt-get update && sudo apt-get install python3-dev python3-pip python3-venv python-virtualenv
```

<br>

#### **1.0.2 Create Virtual Environgment (Recommended)**


___

**Create virtual environment**

>On a terminal, or using magic keys, create a new virtual environment using a Python interpreter and creating a ```venv``` directory to hold it

```bash
    $ python3 -m venv venv
```

<br>

**Activate the virtual environment**

```bash
    $ source ./venv/bin/activate      # sh, bash, or zsh

    $ . ./venv/bin/activate.fish       # fish

    $ source ./venv/bin/activate.csh  # csh or tcsh
```

<br>

**Install packages within virtual environment**

When the virtual environment is active, your shell prompt is prefixed by ```(venv)```

> Step 1: reapeat 1.1.1 above, within ```(venv)```<br>

```bash
    # TensorFlow requires pip version >= 19.0
    $ pip install --upgrade pip 
```    

> Step 2: Install TensorFlow2

```bash
    (venv) $ pip install --upgrade tensorflow
```


> Step 3: Verify install

```bash
    (venv) $ python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])))"
```

> Step 4: To exit virtual enviroment later

```bash
    (venv) $ pip deactivate
```

<br>

#### **1.0.3 Install TensorFlow on Local Machine (Optional Setup)**

___


**TensorFlow 2**

https://www.tensorflow.org/install/pip

```bash
    # Install using pip
    $ pip install --upgrade tensorflow 
```

<br>

**Verify Installation**

```bash
    $ python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])))"
```

<br>

#### **1.0.4 Run TensorFlow in a Container (Optional Setup)**

___

[**Tensorflow Docker images**](https://www.tensorflow.org/install/docker) come pre-configured to run TensorFlow, and provide a virtual environment that is generally the easiest way to set up [**GPU processing**](https://www.tensorflow.org/install/gpu). To learn more about how to use [**Docker**](https://www.tensorflow.org/install/docker/) to separate your applications from your infrastructure, click [here](https://docs.docker.com/get-docker/).

<br>

<a id='Part_1_1'></a>

<hr style="border: 1px solid#003262;" />

#### PART 1.1: TENSORFLOW SETUP

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/tf_logo_social.png" align="center" width="30%" padding="0px"><br>
    <br>
</div>

<br> 

#### **1.1.1 General Notebook Setup**
___

In [1]:
# Pyton 2 and 3 support
from __future__ import division, print_function, unicode_literals

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

# Hide warnings
import warnings
warnings.filterwarnings('ignore')

<br>

**Install TensorFlow:** the following will install TensorFlow using bash. It serves our purposes. However, for detailed installation instructions see [Part 1.1](#Part_1_1)

In [2]:
## Create Virtual Environment ##
! python3 -m venv ./assets/venv

In [3]:
## Activate Virtual Environment ##
! . ./assets/venv/bin/activate

In [4]:
## Ensure pip version >= 19.0 ##
! pip install --upgrade pip 

Requirement already up-to-date: pip in /home/ehch/anaconda3/lib/python3.6/site-packages (20.1.1)


In [5]:
## Ensure TF version >= 2.0 ##
! pip install --upgrade tensorflow

<br>

**Import TensorFlow**

In [6]:
# Canonical way of importing TensorFlow
import tensorflow as tf

___

**Note:** If ```import tensorflow as tf``` doesn't work, TensorFlow is not installed correctly. To resolve the issue see [**build and install error messages**](https://www.tensorflow.org/install/errors)
    
___

In [7]:
# Check tf version, oftentimes tensorflow is not backwards compatible
tf.__version__

'2.2.0'

<a id='Part_1_2'></a>

<hr style="border: 1px solid#003262;" />

#### PART 1.2: TENSORBOARD SETUP

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/tensorboard_logo_social.png" align="center" width="30%" padding="0px"><br>
    <br>
</div>

<br>

[**TensorBoard**](https://www.tensorflow.org/tensorboard/get_started) is a tool for measuring and visualizing machine learning workflows. It enables the tracking of important metrics such as [**acuracy**](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Accuracy), it can display summary operations, tensor ouputs, project embeddings to lower dimensional space, and much more. This broad functionality makes TensorBoard an important tool in evaluating, optimizing, and debugging machine learning models. For a five-minute summary of TensorBoard, click [here](https://youtu.be/3bownM3L5zM).

<br>

#### **1.2.1 Load TensorBoard and Associated Libraries**

___

**Load TensorBoard**

In [18]:
## Load TensorBoard notebook extension ##
%load_ext tensorboard

In [19]:
## Load additional libraries needed ##
from datetime import datetime
from IPython.display import clear_output, Image, display, HTML
import os
import pathlib

t = datetime.utcnow().strftime("%Y%m%d%H%M%S") 
log_dir = "tf_logs"
logd = "/tmp/{}/r{}/".format(log_dir, t)

**Clear TensorBoard Logs**

In [20]:
## Clear any logs from previous runs ##
!rm -rf ./logs/

<br>

#### **1.2.2 Confirm TensorBoard is running**

**Note:** The purpose of this section is only to highlight some of the commands for using TensorBoard. The [**MNIST**](https://en.wikipedia.org/wiki/MNIST_database) data set used here, is only for the purposes of highlighting functionality which on this case depends on the [**Keras**](https://keras.io/) library.

___

In [21]:
# from Keras datasets import mnist
mnist = tf.keras.datasets.mnist

# load data into training and testing splits
(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

# create a sequential model
def create_model():
  return tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation='softmax')
  ])

___

**Note:** Click on the following to learn more about the command:
>[```keras.model.sequential( )```](https://keras.io/api/models/sequential/)<br>
>[```keras.layers.Flatten( )```](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Flatten)<br>
>[```keras.layers.Dense( ) ```](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense)<br>
>[```keras.layers.Dropout( )```](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dropout)
    
___

**Add Tensorboard callbacks**

In [22]:
# instantiate model class
model = create_model()

# compile model 
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# place logs in a timestamped subdirectory and enable histogram computation with every epoch
log_dir = "logs/fit/" + datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)


# Fit model
model.fit(x=x_train, 
          y=y_train, 
          epochs=3, 
          validation_data=(x_test, y_test), 
          # Adding 'callbacks=[tensorboard_callback]' to model.fit( ) ensures that logs are created and stored
          callbacks=[tensorboard_callback])

Epoch 1/3
1875/1875 [==============================] - 8s 5ms/step - loss: 0.2181 - accuracy: 0.9359 - val_loss: 0.1041 - val_accuracy: 0.9688
Epoch 2/3
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0978 - accuracy: 0.9699 - val_loss: 0.0831 - val_accuracy: 0.9750
Epoch 3/3
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0701 - accuracy: 0.9782 - val_loss: 0.0726 - val_accuracy: 0.9777


___

**Note:** Click on the following to learn more about the command:
>[```model.compile( )```](https://keras.io/api/models/model/)<br>
>[```callbacks.TensorBoard( )```](https://keras.io/api/callbacks/)<br>
>[```model.fit( ) ```](https://keras.io/api/models/model_training_apis/)
    
___

**Display Tensorboard Logs**

In [23]:
## %tensorboard line magic. In command line run same command without "%" ##
%tensorboard --logdir logs/fit

Reusing TensorBoard on port 6006 (pid 4823), started 0:09:15 ago. (Use '!kill 4823' to kill it.)

___
**Note:** if TensorBoard does not load after following the above command, try reloading TensordBoard ```%load_ext tensorboard``` 
___

<a id='Part_1_3'></a>

<hr style="border: 1px solid#003262;" />

#### PART 1.3: TENSORFLOW TENSORS

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/DELETE_scalar_to_tensor.png" align="center" width="50%" padding="10px"><br>
    <br>
</div>

[**Tensors**](https://www.tensorflow.org/guide/tensor) are immutable multidimensional arrays for working with data with more than 2-Dimensions. Tensors essentially multilinear maps from vector spaces to real numbers. Hence a tensor can be used to represent scalars, vectors, and matrices. Moreover, tensors are [**highly efficient**], can be ran on several architectures (CPUs, GPUs, Mobile, and Distributed), and make the calculation of gradients easier -- this is particulary useful as the analytical solution of gradients are extremely tedious to derive.

#### **1.2.0 Load Packages/Libraries**

___

#### **1.2.1 Create Dataframe Using an Array**

___

[1.76405235 0.40015721 0.97873798]
[ 2.2408932   1.86755799 -0.97727788]
[ 0.95008842 -0.15135721 -0.10321885]


0
0  1.764052
1  0.400157
2  0.978738

#### **1.2.2 Create Dataframe Using a Dictionary**

___

{'a1': array([1.76405235, 0.40015721, 0.97873798]),
 'a2': array([ 2.2408932 ,  1.86755799, -0.97727788]),
 'a3': array([ 0.95008842, -0.15135721, -0.10321885])}

<!--Navigate back to table of contents-->
<div align="left" style="text-align: left; background-color:#003262;">
    <span>
        <hr style="border: 8px solid#003262;" />
        <a style="color:#FFFFFF; background-color:#003262; border:1px solid #FFFFFF; border-color:#FFFFFF;border-radius:0px;border-width:0px;display:inline-block;font-family:arial,helvetica,sans-serif;font-size:24px;letter-spacing:0px;line-height:20px;padding:24px 40px;text-align:left;text-decoration:none; align:left"> 
            <strong>CONCEPT</strong> CHECK 
        </a>        
    </span>

</div>
<!-------------------------------------->

> **It is easy to add new rows and columns of data to a Pandas DataFrame, and the data can be of any native type (i.e. interger, float, list, series, etc.)  However, much like in dictionaries, dimensions must match.** 

<br>

```python
# Create a new column in the df1 dataframe named 'L' and add the following entries
("List", 3), "3", ["pi", 3.1415]
```
>**The result should be a** ```3x5 matrix```. **What is the output?**

<hr style="border: 2px solid#003262;" />

<a id='Part_1_4'></a>

<hr style="border: 1px solid#003262;" />

#### PART 1.4: TENSORFLOW OPERATIONS
<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/DELETE_tensorflow_operation.png" align="center" width="40%" padding="0px"><br>
    <br>
</div>

#### **1.3.1 Basic Data Selection and Manipulation**

___

a1        a2        a3         k             L  Series
1  1.764052  2.240893  0.950088       3.2     (List, 3)     5.0
2  0.400157  1.867558 -0.151357      True             3     3.0
3  0.978738 -0.977278 -0.103219  [1, one]  [pi, 3.1415]     NaN

<br>

#### **1.3.2 Data Selection: Indexing and Slicing a DataFrame**

>[Indexing DataFrames](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#different-choices-for-indexing)<br>
>[Range Slicing DataFrames](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#slicing-ranges)

___

**Attribute Access**

1    1.764052
2    0.400157
3    0.978738
Name: a1, dtype: float64

<!--Navigate back to table of contents-->
<br>
<div alig="right" style="text-align: right">
    <span>
        <a style="color:#FFFFFF; background-color:#003262; border:1px solid #FFFFFF; border-color:#FFFFFF;border-radius:5px;border-width:0px;display:inline-block;font-family:arial,helvetica,sans-serif;font-size:10px;letter-spacing:0px;line-height:10px;padding:10px 20px;text-align:center;text-decoration:none; align:center" href="#Part_table_contents" name="Table of Contents"  id="Part_table_contents"> 
            Table of Contents 
        </a>
    </span>
</div>
<!-------------------------------------->

<a id='Part_1_5'></a>

<hr style="border: 1px solid#003262;" />

#### PART 1.5: TENSORFLOW COMPUTATION GRAPHS
<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/DELETE_tensorflow_computation_graph.png" align="center" width="40%" padding="0px"><br>
    <br>
</div>

<!--Navigate back to table of contents-->
<div alig="right" style="text-align: right">
    <span>
        <a style="color:#FFFFFF; background-color:#003262; border:1px solid #FFFFFF; border-color:#FFFFFF;border-radius:5px;border-width:0px;display:inline-block;font-family:arial,helvetica,sans-serif;font-size:10px;letter-spacing:0px;line-height:10px;padding:10px 20px;text-align:center;text-decoration:none; align:center" href="#Part_table_contents" name="Table of Contents"  id="Part_table_contents"> 
            Table of Contents 
        </a>
    </span>
</div>
<!-------------------------------------->

<a id='Part_2'></a>

<hr style="border: 2px solid#003262;" />

#### PART 2

## **TENSORFLOW** GRAPHS, **AND** EXECUTIONS

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/pandas_logo.png" align="center" width="15%" padding="10"><br>
    <br>
</div>

<br>

[**TODO**](https://pandas.pydata.org/pandas-docs/stable/index.html) TODO

#### CONTENTS:

> [PART 2.1: TENSORFLOW OPERATIONS](#Part_2_1)<br>
> [PART 2.2: TO DO](#Part_2_2)<br>


<a id='Part_2_1'></a>

<hr style="border: 1px solid#003262;" />

#### PART 2.1:TENSORFLOW OPERATIONS

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/pandas-16.png" align="center" width="40%" padding="0px"><br>
    <br>
</div>

<br>

#### **2.1.0 Load Packages/Libraries**

___

<br>

#### **2.2.1 Read Data Into DataFrame From CSV File**


___

<br>

#### **2.2.2 Explore the Data**

___

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Admit   24 non-null     object
 1   Gender  24 non-null     object
 2   Dept    24 non-null     object
 3   Freq    24 non-null     int64 
dtypes: int64(1), object(3)
memory usage: 896.0+ bytes


<a id='Part_2_2'></a>

<hr style="border: 1px solid#003262;" />

#### PART 2.2: TO DO

<br>

<!--Navigate back to table of contents-->
<div alig="right" style="text-align: right">
    <span>
        <a style="color:#FFFFFF; background-color:#003262; border:1px solid #FFFFFF; border-color:#FFFFFF;border-radius:5px;border-width:0px;display:inline-block;font-family:arial,helvetica,sans-serif;font-size:10px;letter-spacing:0px;line-height:10px;padding:10px 20px;text-align:center;text-decoration:none; align:center" href="#Part_table_contents" name="Table of Contents"  id="Part_table_contents"> 
            Table of Contents 
        </a>
    </span>
</div>
<!-------------------------------------->

<a id='Part_3'></a>

<hr style="border: 2px solid#003262;" />

#### PART 3 (OPTIONAL)

## **TENSORFLOW** LINEAR **REGRESSION**

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/pandas_logo.png" align="center" width="15%" padding="10"><br>
    <br>
</div>

<br>

[**TODO**](https://pandas.pydata.org/pandas-docs/stable/index.html) TODO

#### CONTENTS:

> [PART 3.1: CONSTRUCTION PHASE](#Part_3_1)<br>
> [PART 3.2: EVALUATION PHASE](#Part_3_2)<br>
> [PART 3.3: ACCURACY COMPARISON](#Part_3_3)

<a id='Part_3_1'></a>

<hr style="border: 1px solid#003262;" />

#### PART 3.1: CONSTRUCTION PHASE

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/pandas-16.png" align="center" width="40%" padding="0px"><br>
    <br>
</div>

<br>

<a id='Part_3_2'></a>

<hr style="border: 1px solid#003262;" />

#### PART 3.2: EVALUATION PHASE

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/pandas-16.png" align="center" width="40%" padding="0px"><br>
    <br>
</div>

<br>

<a id='Part_3_3'></a>

<hr style="border: 1px solid#003262;" />

#### PART 3.3: ACCURACY COMPARISON

<br>

<div align="center" style="font-size:12px; font-family:FreeMono; font-weight: 100; font-stretch:ultra-condensed; line-height: 1.0; color:#2A2C2B">
    <img src="/assets/content/images/pandas-16.png" align="center" width="40%" padding="0px"><br>
    <br>
</div>

<br>

<!--Navigate back to table of contents-->
<div alig="right" style="text-align: right">
    <span>
        <a style="color:#FFFFFF; background-color:#003262; border:1px solid #FFFFFF; border-color:#FFFFFF;border-radius:5px;border-width:0px;display:inline-block;font-family:arial,helvetica,sans-serif;font-size:10px;letter-spacing:0px;line-height:10px;padding:10px 20px;text-align:center;text-decoration:none; align:center" href="#Part_table_contents" name="Table of Contents"  id="Part_table_contents"> 
            Table of Contents 
        </a>
    </span>
</div>
<!-------------------------------------->

<a id='Part_4'></a>

<hr style="border: 2px solid#003262;" />

#### PART 4

## **WRAP-UP** AND **NEXT** STEPS


<div align="center">
    <img src= "/assets/content/datax_logos/DataX_icon_wide_logo.png" align="center" width="80%" padding="20">
</div>

<br>

As you may have started to notice, that there is much more than can be done using Flask. Wanting to learn other uses of Flask?  Visit the [**Data-X website**](https://datax.berkeley.edu/) to learn more, or use the following links to topics of interest:

> [TODO (m---): url needed]() TODO TODO TODO <br>
> [TODO (m---): url needed]() TODO TODO TODO <br>
> [TODO (m---): url needed]() TODO TODO TODO



<br>

<br>


<a id='Appendix_1'></a>

<hr style="border: 6px solid#003262;" />


#### APPENDIX I


## PREREQUISITE **KNOWLEDGE** AND **REQUIREMENTS**


[**TODO**](https://pandas.pydata.org/about/) TODO TODO TODO

##### **PYTHON**
This notebook and the executable files are built using [**Python**](https://www.python.org/) and relies on common Python packages (e.g. [**NumPy**](https://numpy.org/)) for operation. If you have a lot of programming experience in a different language (e.g. C/C++/Matlab/Java/Javascript), you will likely be fine, otherwise:

> [**Python (EDX free)**](https://www.edx.org/course?search_query=python)<br>
> [**Python (Coursera free)**](https://www.coursera.org/search?query=python)


##### **NUMPY**
This notebook and the executable files are built using [**Python**](https://www.python.org/) and relies on common Python packages (e.g. [**NumPy**](https://numpy.org/)) for operation. If you have a lot of programming experience in a different language (e.g. C/C++/Matlab/Java/Javascript), you will likely be fine, otherwise:

> [**Python (EDX free)**](https://www.edx.org/course?search_query=python)<br>
> [**Python (Coursera free)**](https://www.coursera.org/search?query=python)


##### **PANDAS**
This notebook and the executable files are built using [**Python**](https://www.python.org/) and relies on common Python packages (e.g. [**NumPy**](https://numpy.org/)) for operation. If you have a lot of programming experience in a different language (e.g. C/C++/Matlab/Java/Javascript), you will likely be fine, otherwise:

> [**Python (EDX free)**](https://www.edx.org/course?search_query=python)<br>
> [**Python (Coursera free)**](https://www.coursera.org/search?query=python)


##### **MACHINE LEARNING PROBLEM FRAMING**
This notebook and the executable files are built using [**Python**](https://www.python.org/) and relies on common Python packages (e.g. [**NumPy**](https://numpy.org/)) for operation. If you have a lot of programming experience in a different language (e.g. C/C++/Matlab/Java/Javascript), you will likely be fine, otherwise:

> [**Python (EDX free)**](https://www.edx.org/course?search_query=python)<br>
> [**Python (Coursera free)**](https://www.coursera.org/search?query=python)


##### **BASH**
This notebook and the executable files are built using [**Python**](https://www.python.org/) and relies on common Python packages (e.g. [**NumPy**](https://numpy.org/)) for operation. If you have a lot of programming experience in a different language (e.g. C/C++/Matlab/Java/Javascript), you will likely be fine, otherwise:

> [**Python (EDX free)**](https://www.edx.org/course?search_query=python)<br>
> [**Python (Coursera free)**](https://www.coursera.org/search?query=python)

<a id='Appendix_2'></a>

<hr style="border: 2px solid#003262;" />

#### APPENDIX II


## **REFERENCES** AND ADDITIONAL **RESOURCES**

<br>


> * [**TensorFlow Tutorials**](https://pandas.pydata.org/pandas-docs/stable/user_guide/index.html)

> * [**TensorFlow Guide**](https://youtu.be/dcqPhpY7tWk) 

> * [**Introduction to TensorFlow by Andrew Ng and Kian Katanforoosh**](https://cs230.stanford.edu/blog/tensorflow/)

> * [**Deep Learning Illustrated by Jon Krohn, Grant Beyleveld, and Aglae Bassens**](https://www.deeplearningillustrated.com/)

> * [**Hands-On Introduction to TensorFlow 2.0 by Josh Gordon and Amit Patankar**](https://youtu.be/Yyv-ng0_OTU)

> * [**Getting Started with TensorFlow and Deep Learning by Josh Gordon**](https://youtu.be/tYYVSEHq-io)

<!--Navigate back to table of contents-->
<div alig="right" style="text-align: right">
    <span>
        <a style="color:#FFFFFF; background-color:#003262; border:1px solid #FFFFFF; border-color:#FFFFFF;border-radius:5px;border-width:0px;display:inline-block;font-family:arial,helvetica,sans-serif;font-size:10px;letter-spacing:0px;line-height:10px;padding:10px 20px;text-align:center;text-decoration:none; align:center" href="#Part_table_contents" name="Table of Contents"  id="Part_table_contents"> 
            Table of Contents 
        </a>
    </span>
</div>
<!-------------------------------------->

<hr style="border: 6px solid#003262;" />
